src/transformers/models/gpt_neox/modeling_gpt_neox.py

```diff
@@ -297,21 +297,22 @@ def __init__(self, dim, max_position_embeddings, base=10000, device=None):
         self.dim = dim
         self.max_position_embeddings = max_position_embeddings
         self.base = base
-        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
+        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2).float() / self.dim))
+        inv_freq = inv_freq.to(device, dtype=torch.get_default_dtype())
         self.register_buffer("inv_freq", inv_freq)
 
         # Build here to make `torch.jit.trace` work.
         self._set_cos_sin_cache(seq_len=max_position_embeddings, device=self.inv_freq.device)
 
     def _set_cos_sin_cache(self, seq_len, device):
         self.max_seq_len_cached = seq_len
-        t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)
+        t = torch.arange(self.max_seq_len_cached, device=device).float()
```

Comment:
Since emb.cos() and emb.sin() at line 314 can only be calculated float32 on CPU, this variable must be float32.

If this `t` is float16 and emb.cos() is calculated on CPU, the following error will be raised:
```
RuntimeError: "cos_vml_cpu" not implemented for 'Half'
```